### PR TITLE
Remember state between sessions

### PR DESCRIPTION
**Pull request by humphd**
This uses storage on in the hosting app to keep track of UI and project state.  It then sends it to Bramble on startup.

It doesn't currently play nicely with all of our Thimble UI, since we seem to assume startup values instead of reading them from initial state.  It won't be hard to fix, though.

### CODE REVIEW

File: src/bramble/client/StateManager.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        return window.localStorage;
+    }(window));
+
+    function prefix(name) {
+        return "bramble-property::" + name;
+    }
+
+    function getString(storage, property) {
+        var val = storage.getItem(prefix(property));
+        return val || null;
```
Comment (sedge) [inline src/bramble/client/StateManager.js]: Why `null`?

Comment (humphd) [inline src/bramble/client/StateManager.js]: It just makes what `localStorage.getItem()` does when the key isn't there more explicit (i.e., it returns null, but this documents the fact that `getString` can return null).

**Comment by humphd** [comment]
I've rebased this onto bramble, and it's ready for review.

File: src/bramble/client/StateManager.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            return false;
+        } else {
+            // not set
+            return null;
+        }
+    }
+
+    function getInt(storage, property) {
+        var str = getString(storage, property);
+        return str|0;
```
Comment (sedge) [inline src/bramble/client/StateManager.js]: I've seen you use this trick before. How does it work? It just seems to return `0` if `str` doesn't exist, so is it a performance thing over `str || 0`?

File: src/bramble/client/main.js [hunk 3]
```diff
@@ -113,10 +114,8 @@ define([
 
     // After calling Bramble.load(), Bramble.mount() specifies project root entry path info
     Bramble.mount = function(root, filename) {
-        // Assume index.html if no filename provided.
         if (!filename) {
-            filename = "index.html";
```
Comment (sedge) [inline src/bramble/client/main.js]: These seems like a useful default. Why are we removing it?

File: src/bramble/client/main.js [hunk 4]
```diff
@@ -318,6 +313,14 @@ define([
             function _mount() {
                 setReadyState(Bramble.MOUNTING);
 
+                root = Path.normalize(root);
+                filename = filename || (_state.filename || "index.html");
```
Comment (sedge) [inline src/bramble/client/main.js]: Ah, gotcha.

File: src/bramble/client/main.js (hunk 3 above)
Comment (sedge) [inline src/bramble/client/main.js]: Nevermind, I see now that you moved it.

File: src/bramble/client/StateManager.js (hunk 2 above)
Comment (humphd) [inline src/bramble/client/StateManager.js]: It casts the value to a 32-bit Int.  It's like saying `parseInt(str)`, but shorter :)  Doing `str || 0` is not the same thing, even though they both return 0 if `str` is null.

```
> 3.14|0
3
> 3.14||0
3.14
```

Comment (sedge) [inline src/bramble/client/StateManager.js]: Ahhh, very cool. Thanks!

**Comment by humphd** [comment]
r+ from @gideonthomas on irc.

**Pull request merged**